File: 536. Construct Binary Tree From String/iterative.py
class Solution(object):
    def str2tree(self, s):
        """
        :type s: str
        :rtype: TreeNode
        """
        if not s:
            return None
        stack = []
        i = 0
        while i < len(s):
            c = s[i]
            if c == ')':
                stack.pop()
            elif c == '-' or ord(c) >= ord('0') and ord(c) <= ord('9'):
                while i+1 < len(s) and ord(s[i+1]) <= ord('9') and ord(s[i+1]) >= ord('0'):
                    i += 1
                cur = TreeNode(int(s[j: i+1]))
                if stack:
                    t = stack[-1]
                    if t.left:
                        t.right = cur
                    else:
                        t.left = cur
                stack.append(cur)
            i += 1
        return stack[-1] if stack else None

File: 114. Flatten Binary Tree to Linked List/iterative.java
class Solution {
    public void flatten(TreeNode root) {
        TreeNode now = root;
        while (now != null) {
            if (now.left != null) {
                // FIND THE CURRENT NODE'S PRENODE THAT LINKS TO CURRENT NDOES RIGHT SUBTREE.
                TreeNode prev = now.left;
                while (prev.right != null) {
                    prev = prev.right;
                }
                prev.right = now.right;
                // use current node's left subtree to replace its right subtree(original right).
                // subtree is already linked by current node's prenode
                now.right = now.left;
                now.left = null;
            }
            now = now.right;
        }
    }
}